packages/flutter/lib/src/services/binding.dart

```diff
@@ -186,8 +186,11 @@ class _DefaultBinaryMessenger extends BinaryMessenger {
     ByteData response;
     try {
       final MessageHandler handler = _handlers[channel];
-      if (handler != null)
+      if (handler != null) {
         response = await handler(data);
+      } else {
+        ui.channelBuffers.push(channel, data, callback);
```

Comment:
Since this is an entirely new API that we built and we can make it do anything we want, how about:

Make the dart:ui side always push all messages into the buffer unless that one channel had its buffer drained at least once. i.e., instead of switching on onPlatformMessage being null or not, just always push into the channelBuffers. In the channel buffers, keep a list of channels that's been drained. If drained once, stop queueing more into it. We can rename that thing DrainOnceChannelBuffer or something to make it clear. 

The main advantage would be it being self-contained and not needing orchestration from both sides to function.

Reply:
An interesting idea.  I've thought of something similar where all messages are put into channel buffers as part of normal message passing.  I decided against that since it's kind of wasteful to use channelBuffers when we don't need to.  Communicating via the callstack seems safer.

Responding more directly, it doesn't eliminate orchestration since we still need to drain here.  I think it is a bit harder to follow logically.  Given 2 options I usually prefer the option that results in less state.  Many bugs come from bad stored state because it's harder to reason about when reading code, more stored state means more chances for bugs.

Reply:
Ok SG. 

Let's add some tests on this current mechanism